src/core/qgsproject.cpp

```diff
@@ -1141,20 +1115,15 @@ void QgsProject::cleanTransactionGroups( bool force )
 bool QgsProject::readLayer( const QDomNode& layerNode )
 {
   QList<QDomNode> brokenNodes;
-  QList< QPair< QgsVectorLayer*, QDomElement > > vectorLayerList;
-  if ( addLayer( layerNode.toElement(), brokenNodes, vectorLayerList ) )
+  if ( addLayer( layerNode.toElement(), brokenNodes ) )
   {
     // have to try to update joins for all layers now - a previously added layer may be dependent on this newly
     // added layer for joins
     QVector<QgsVectorLayer*> vectorLayers = layers<QgsVectorLayer*>();
     Q_FOREACH ( QgsVectorLayer* layer, vectorLayers )
     {
-      processLayerJoins( layer );
-    }
-
-    if ( !vectorLayerList.isEmpty() )
-    {
-      emit readMapLayer( vectorLayerList.at( 0 ).first, vectorLayerList.at( 0 ).second );
+      // TODO: should be only done later - and with all layers (other layers may have referenced this layer)
```

Comment:
Yes please!

There's a plugin which does this - can't remember the name offhand... something about change layer source, and it's very handy behaviour